ros/src/system/autoware_health_checker/src/diag_buffer.cpp

```diff
@@ -0,0 +1,93 @@
+#include <autoware_health_checker/diag_buffer.h>
+
+namespace autoware_health_checker
+{
+    DiagBuffer::DiagBuffer(std::string key, uint8_t type, std::string description, double buffer_length) : type(type), description(description)
+    {
+        key_ = key;
+        buffer_length_ = ros::Duration(buffer_length);
+    }
+
+    DiagBuffer::~DiagBuffer()
+    {
+        
+    }
+
+    void DiagBuffer::addDiag(autoware_system_msgs::DiagnosticStatus status)
+    {
+        std::lock_guard<std::mutex> lock(mtx_);
+        buffer_[status.level].status.push_back(status);
+        updateBuffer();
+        return;
+    }
+
+    autoware_system_msgs::DiagnosticStatusArray DiagBuffer::getAndClearData()
+    {
+        autoware_system_msgs::DiagnosticStatusArray data;
+        data = buffer_[autoware_health_checker::LEVEL_FATAL];
+        data.status.insert(data.status.end(),buffer_[autoware_health_checker::LEVEL_ERROR].status.begin(),buffer_[autoware_health_checker::LEVEL_ERROR].status.end());
+        data.status.insert(data.status.end(),buffer_[autoware_health_checker::LEVEL_WARN].status.begin(),buffer_[autoware_health_checker::LEVEL_WARN].status.end());
+        data.status.insert(data.status.end(),buffer_[autoware_health_checker::LEVEL_OK].status.begin(),buffer_[autoware_health_checker::LEVEL_OK].status.end());
+        data.status.insert(data.status.end(),buffer_[autoware_health_checker::LEVEL_UNDEFINED].status.begin(),buffer_[autoware_health_checker::LEVEL_UNDEFINED].status.end());
+        std::sort(data.status.begin(), data.status.end(), std::bind(&DiagBuffer::compareTimestamp, this, std::placeholders::_1, std::placeholders::_2));
+        buffer_.clear();
+        return data;
+    }
+
+    uint8_t DiagBuffer::getErrorLevel()
+    {
+        std::lock_guard<std::mutex> lock(mtx_);
+        updateBuffer();
+        if(buffer_[autoware_health_checker::LEVEL_FATAL].status.size() != 0)
+        {
+            return autoware_health_checker::LEVEL_FATAL;
+        }
+        else if(buffer_[autoware_health_checker::LEVEL_ERROR].status.size() != 0)
+        {
+            return autoware_health_checker::LEVEL_ERROR;
+        }
+        else if(buffer_[autoware_health_checker::LEVEL_WARN].status.size() != 0)
+        {
+            return autoware_health_checker::LEVEL_WARN;
+        }
+        else
+        {
+            return autoware_health_checker::LEVEL_OK;
+        }
+    }
+
+    // filter data from timestamp and level
+    autoware_system_msgs::DiagnosticStatusArray DiagBuffer::filterBuffer(ros::Time now, uint8_t level)
+    {
+        autoware_system_msgs::DiagnosticStatusArray filterd_data;
+        autoware_system_msgs::DiagnosticStatusArray ret;
+        if(buffer_.count(level) != 0)
+        {
+            filterd_data = buffer_[level];
```

Comment:
Does it really seem necessary to copy?
I think that copying all the data seems to be useless.
Is it okay to pass by reference?